### PR TITLE
feat(ngx-layout): WCAG/ARIA for drag-and-drop

### PR DESCRIPTION
**Pull request by IbenTesara**

### CODE REVIEW

**Comment by IbenTesara** [comment]
@ekincia @WHeirstrate @DenisValcke 

Once we all given our okay on the default text, we'll look into adding more languages 👍

File: libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-item.directive.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+ */
+@Directive({
+	selector: '[ngxAccessibleDragAndDropItem]',
+	exportAs: 'ngxAccessibleDragAndDropItem',
+	standalone: true,
+	host: {
+		'[attr.tabIndex]': 'tabIndex()',
+	},
+})
+export class NgxAccessibleDragAndDropItemDirective extends NgxHasFocusDragAndDropDirective {
```
Comment (WHeirstrate) [inline libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-item.directive.ts:31]: I'm gonna be honest, I don't really like the fact that `Accessible` is always visible in the services, directives and types. Yes, we focus on the a11y, but should it be this explicitly mentioned in each name?

**Review by WHeirstrate — APPROVED** [review]

File: libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-item.directive.ts (hunk 1 above)
Comment (IbenTesara) [reply libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-item.directive.ts:31]: I mean, the sole purpose of these directives, types and services is accessibility... They serve no other purpose than that; I don't understand why we'd not include it in the name? 😅 

It's not like we do this with everything we do, this is just 100% intended for accessible drag and drop, I think the intention should be clear here.

File: libs/layout/src/lib/abstracts/drag-and-drop/drag-and-drop.service.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	public setDragAndDropDescription(parent: HTMLElement, description?: string): Observable<void> {
+		// Iben: Create the description element and its id
+		const element: HTMLParagraphElement = document.createElement('p');
+		const id: string = UUID.UUID();
+
+		// Iben: Take the current language to fetch the message
+		return (
+			typeof this.currentLanguage === 'string'
+				? of(this.currentLanguage)
+				: this.currentLanguage
```
Comment (DenisValcke) [inline libs/layout/src/lib/abstracts/drag-and-drop/drag-and-drop.service.ts:121]: What if a non-string value is passed that is not evaluated as falsy?

File: libs/layout/src/lib/abstracts/drag-and-drop/drag-and-drop.service.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	 * @param description - An optional description used to overwrite the default description
+	 */
+	public setDragAndDropDescription(parent: HTMLElement, description?: string): Observable<void> {
+		// Iben: Create the description element and its id
+		const element: HTMLParagraphElement = document.createElement('p');
+		const id: string = UUID.UUID();
+
+		// Iben: Take the current language to fetch the message
+		return (
+			typeof this.currentLanguage === 'string'
```
Comment (DenisValcke) [inline libs/layout/src/lib/abstracts/drag-and-drop/drag-and-drop.service.ts:119]: What if this evaluates as a non-string value, but the value is not an observable (for example `undefined` or something else)?

File: libs/layout/src/lib/components/configurable-layout/configurable-layout.component.ts [hunk 4]
```diff
@@ -341,6 +364,32 @@ export class NgxConfigurableLayoutComponent
...
+		const startArray = formData[event.previousContainer];
+		const endArray = formData[event.newContainer];
+
+		// Iben: If the drag and drop is within the same row, we move
+		if (event.previousContainer === event.newContainer) {
+			moveItemInArray(endArray, event.previousIndex, event.newIndex);
+		} else {
+			// Iben: If the drag and drop is over multiple rows, we transfer
+			transferArrayItem(startArray, endArray, event.previousIndex, event.newIndex);
+			formData[event.previousContainer] = startArray;
```
Comment (DenisValcke) [inline libs/layout/src/lib/components/configurable-layout/configurable-layout.component.ts:384]: Just checking: Is the `previousContainer` always defined?

File: libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-host.directive.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	/**
+	 * Returns the container based on the provided index
+	 *
+	 * @param index - The index of the container
+	 */
+	public getContainer(index: number): NgxAccessibleDragAndDropContainerDirective {
+		return this.containers.find((container) => container.index === index);
+	}
+
+	ngAfterViewInit() {
```
Comment (DenisValcke) [inline libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-host.directive.ts:56]: Member visibility & return type are missing here.

File: libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-item.directive.ts
Comment (DenisValcke) [inline libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-item.directive.ts:1]: Check member visibility of the properties. some are missing.

File: libs/layout/src/lib/directives/drag-and-drop/has-focus.directive.ts
Comment (DenisValcke) [inline libs/layout/src/lib/directives/drag-and-drop/has-focus.directive.ts:1]: Same not here as the previous one.

Comment (DenisValcke) [inline libs/layout/src/lib/directives/drag-and-drop/has-focus.directive.ts:1]: This directive is an abstract, but the classname does not reflect that. Is this intentional?

File: libs/layout/src/lib/services/live-region/live-region.service.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+		this.region.setAttribute('aria-live', 'assertive');
+		this.region.setAttribute('role', 'log');
+		this.region.setAttribute('aria-relevant', 'additions');
+		this.region.setAttribute('aria-atomic', 'true');
+
+		//Iben: Move the live region out of sight for without assistive technologies
+		hideElement(this.region);
+
+		// Iben: Append the live region to the body
+		document.body.appendChild(this.region);
```
Comment (DenisValcke) [inline libs/layout/src/lib/services/live-region/live-region.service.ts:69]: Same here.

File: libs/layout/src/lib/components/configurable-layout/configurable-layout.component.html [hunk 7]
```diff
@@ -1,55 +1,63 @@
...
+		[ngxAccessibleDragAndDropContainerIndex]="index"
+		[ngxAccessibleDragAndDropContainerLabel]="rowLabel ? rowLabel + ' ' + (index + 1) : ''"
+		[id]="'ngx-layout-row-' + index"
+		[ngStyle]="{ keys: form.value, showInactive } | ngxConfigurableLayoutItemSize : itemSize"
+		[style.column-gap]="columnGap"
+		[cdkDropListSortPredicate]="beforeSort.bind(this)"
+		[cdkDropListEnterPredicate]="beforeDrop.bind(this)"
+		(cdkDropListDropped)="drop($event)"
+	>
+		@for (item of row; track item; let index = $index) { @if (itemTemplateRecord()[item.key]) {
```
[the remark this answers is not in the document]
Comment (IbenTesara) [reply libs/layout/src/lib/components/configurable-layout/configurable-layout.component.html:26]: Seems like a thing we need to tackle the moment we bring in all the other packages?

File: libs/layout/src/lib/abstracts/drag-and-drop/drag-and-drop.service.ts (hunk 3 above)
Comment (IbenTesara) [reply libs/layout/src/lib/abstracts/drag-and-drop/drag-and-drop.service.ts:119]: I get what you mean here, and I agree on checking for null/undefined; but the typing will already prevent you from returning anything but a string or an Observable. If you decide to bypass this by casting to any then that's on you 😂 

`abstract get currentLanguage(): string | Observable<string>;`

File: libs/layout/src/lib/components/configurable-layout/configurable-layout.component.ts (hunk 4 above)
Comment (IbenTesara) [reply libs/layout/src/lib/components/configurable-layout/configurable-layout.component.ts:384]: It should be, but I'll double check here as well

File: libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-host.directive.ts (hunk 5 above)
Comment (IbenTesara) [reply libs/layout/src/lib/directives/drag-and-drop/drag-and-drop-host.directive.ts:56]: Is there an EsLint rule for that somewhere? Might be good to add it to the list ;P

File: libs/layout/src/lib/const/drag-and-drop/drag-and-drop-message.const.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+		reordered: '{{#item}} wurde von Position {{#from}} nach Position {{#to}} verschoben.',
+		moved: '{{#item}} wurde von {{#from}} nach {{#to}} verschoben.',
+		cancelled: 'Das Verschieben von {{#item}} wurde abgebrochen.',
+		item: 'Eintrag',
+		container: 'Liste',
+
+		description:
+			'Um Einträge in diesem Teil zu verschieben, navigiere mit der Tab-Taste zu einem Eintrag. Mit Drücken der Eingabe- oder Leerzeichentaste wählst du einen Eintrag aus. Mit den Pfeiltasten verschiebst du den Eintrag.',
+	},
+	ku: {
```
Comment (DenisValcke) [inline libs/layout/src/lib/const/drag-and-drop/drag-and-drop-message.const.ts:85]: Just a reminder not to merge this until @ekincia validates these translations.

File: libs/layout/src/lib/const/drag-and-drop/drag-and-drop-message.const.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+		deselected: '{{#item}} hate jêbirin.',
+		reordered: '{{#item}} ji cîhê {{#from}} ve ji cîhê {{#to}} ve veguhast.',
+		moved: '{{#item}} hate veguhastin ji {{#from}} ve {{#to}}.',
+		cancelled: 'Veguhastina {{#item}} hate betal kirin.',
+		item: 'şanî',
+		container: 'lista',
+		description:
+			'Ji bo veguhastina şanîyan di vê beşê de, bi bikaranîna tuşeya Tab ve her şanîyekê biguherîne. Ji bo hilbijartina şanîyekê, bişkojka Enter an Space bikirtînin. Bi bikaranîna tuşeya şilî şanîyê biguherînin.',
+	},
+	ru: {
```
Comment (DenisValcke) [inline libs/layout/src/lib/const/drag-and-drop/drag-and-drop-message.const.ts:96]: Just a reminder not to merge this until @SergueySHD validates these translations.

**Pull request merged**